openssh-7.4p1/ssh-keysign.8
.\" $OpenBSD: ssh-keysign.8,v 1.15 2016/02/17 07:38:19 jmc Exp $
.\"
.\" Copyright (c) 2002 Markus Friedl.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate: February 17 2016 $
.Dt SSH-KEYSIGN 8
.Os
.Sh NAME
.Nm ssh-keysign
.Nd ssh helper program for host-based authentication
.Sh SYNOPSIS
.Nm
.Sh DESCRIPTION
.Nm
is used by
.Xr ssh 1
to access the local host keys and generate the digital signature
required during host-based authentication.
.Pp
.Nm
is disabled by default and can only be enabled in the
global client configuration file
.Pa /etc/ssh/ssh_config
by setting
.Cm EnableSSHKeysign
to
.Dq yes .
.Pp
.Nm
is not intended to be invoked by the user, but from
.Xr ssh 1 .
See
.Xr ssh 1
and
.Xr sshd 8
for more information about host-based authentication.
.Sh FILES
.Bl -tag -width Ds -compact
.It Pa /etc/ssh/ssh_config
Controls whether
.Nm
is enabled.
.Pp
.It Pa /etc/ssh/ssh_host_dsa_key
.It Pa /etc/ssh/ssh_host_ecdsa_key
.It Pa /etc/ssh/ssh_host_ed25519_key
.It Pa /etc/ssh/ssh_host_rsa_key
These files contain the private parts of the host keys used to
generate the digital signature.
They should be owned by root, readable only by root, and not
accessible to others.
Since they are readable only by root,
.Nm
must be set-uid root if host-based authentication is used.
.Pp
.It Pa /etc/ssh/ssh_host_dsa_key-cert.pub
.It Pa /etc/ssh/ssh_host_ecdsa_key-cert.pub
.It Pa /etc/ssh/ssh_host_ed25519_key-cert.pub
.It Pa /etc/ssh/ssh_host_rsa_key-cert.pub
If these files exist they are assumed to contain public certificate
information corresponding with the private keys above.
.El
.Sh SEE ALSO
.Xr ssh 1 ,
.Xr ssh-keygen 1 ,
.Xr ssh_config 5 ,
.Xr sshd 8
.Sh HISTORY
.Nm
first appeared in
.Ox 3.2 .
.Sh AUTHORS
.An Markus Friedl Aq Mt markus@openbsd.org


openssh-7.4p1/ChangeLog
commit 4a354fc231174901f2629437c2a6e924a2dd6772
Author: Damien Miller <djm@mindrot.org>
Date:   Mon Dec 19 15:59:26 2016 +1100

    crank version numbers for release

commit 5f8d0bb8413d4d909cc7aa3c616fb0538224c3c9
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Mon Dec 19 04:55:51 2016 +0000

    upstream commit
    
    openssh-7.4
    
    Upstream-ID: 1ee404adba6bbe10ae9277cbae3a94abe2867b79

commit 3a8213ea0ed843523e34e55ab9c852332bab4c7b
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Mon Dec 19 04:55:18 2016 +0000

    upstream commit
    
    remove testcase that depends on exact output and
    behaviour of snprintf(..., "%s", NULL)
    
    Upstream-Regress-ID: cab4288531766bd9593cb556613b91a2eeefb56f

commit eae735a82d759054f6ec7b4e887fb7a5692c66d7
Author: dtucker@openbsd.org <dtucker@openbsd.org>
Date:   Mon Dec 19 03:32:57 2016 +0000

    upstream commit
    
    Use LOGNAME to get current user and fall back to whoami if
    not set. Mainly to benefit -portable since some platforms don't have whoami.
    
    Upstream-Regress-ID: e3a16b7836a3ae24dc8f8a4e43fdf8127a60bdfa

commit 0d2f88428487518eea60602bd593989013831dcf
Author: dtucker@openbsd.org <dtucker@openbsd.org>
Date:   Fri Dec 16 03:51:19 2016 +0000

    upstream commit
    
    Add regression test for AllowUsers and DenyUsers.  Patch from
    Zev Weiss <zev at bewilderbeest.net>
    
    Upstream-Regress-ID: 8f1aac24d52728398871dac14ad26ea38b533fb9

commit 3bc8180a008929f6fe98af4a56fb37d04444b417
Author: Darren Tucker <dtucker@zip.com.au>
Date:   Fri Dec 16 15:02:24 2016 +1100

    Add missing monitor.h include.
    
    Fixes warning pointed out by Zev Weiss <zev at bewilderbeest.net>

commit 410681f9015d76cc7b137dd90dac897f673244a0
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Fri Dec 16 02:48:55 2016 +0000

    upstream commit
    
    revert to rev1.2; the new bits in this test depend on changes
    to ssh that aren't yet committed
    
    Upstream-Regress-ID: 828ffc2c7afcf65d50ff2cf3dfc47a073ad39123

commit 2f2ffa4fbe4b671bbffa0611f15ba44cff64d58e
Author: dtucker@openbsd.org <dtucker@openbsd.org>
Date:   Fri Dec 16 01:06:27 2016 +0000

    upstream commit
    
    Move the "stop sshd" code into its own helper function.
    Patch from Zev Weiss <zev at bewilderbeest.net>, ok djm@
    
    Upstream-Regress-ID: a113dea77df5bd97fb4633ea31f3d72dbe356329

commit e15e7152331e3976b35475fd4e9c72897ad0f074
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Fri Dec 16 01:01:07 2016 +0000

    upstream commit
    
    regression test for certificates along with private key
    with no public half. bz#2617, mostly from Adam Eijdenberg
    
    Upstream-Regress-ID: 2e74dc2c726f4dc839609b3ce045466b69f01115

commit 9a70ec085faf6e55db311cd1a329f1a35ad2a500
Author: dtucker@openbsd.org <dtucker@openbsd.org>
Date:   Thu Dec 15 23:50:37 2016 +0000

    upstream commit
    
    Use $SUDO to read pidfile in case root's umask is
    restricted.  From portable.
    
    Upstream-Regress-ID: f6b1c7ffbc5a0dfb7d430adb2883344899174a98

commit fe06b68f824f8f55670442fb31f2c03526dd326c
Author: dtucker@openbsd.org <dtucker@openbsd.org>
Date:   Thu Dec 15 21:29:05 2016 +0000

    upstream commit
    
    Add missing braces in DenyUsers code.  Patch from zev at
    bewilderbeest.net, ok deraadt@
    
    Upstream-ID: d747ace338dcf943b077925f90f85f789714b54e

commit dcc7d74242a574fd5c4afbb4224795b1644321e7
Author: dtucker@openbsd.org <dtucker@openbsd.org>
Date:   Thu Dec 15 21:20:41 2016 +0000

    upstream commit
    
    Fix text in error message.  Patch from zev at
    bewilderbeest.net.
    
    Upstream-ID: deb0486e175e7282f98f9a15035d76c55c84f7f6

commit b737e4d7433577403a31cff6614f6a1b0b5e22f4
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Wed Dec 14 00:36:34 2016 +0000

    upstream commit
    
    disable Unix-domain socket forwarding when privsep is
    disabled
    
    Upstream-ID: ab61516ae0faadad407857808517efa900a0d6d0

commit 08a1e7014d65c5b59416a0e138c1f73f417496eb
Author: djm@openbsd.org <djm@openbsd.org>
Date:   Fri Dec 9 03:04:29 2016 +0000

    upstream commit
    
    log connections dropped in excess of MaxStartups at
    verbose LogLevel; bz#2613 based on diff from Tomas Kuthan; ok dtucker@
    
    Upstream-ID: 703ae690dbf9b56620a6018f8a3b2389ce76d92b

commit 10e290ec00964b2bf70faab15a10a5574bb80527
Author: Darren Tucker <dtucker@zip.com.au>
Date:   Tue Dec 13 13:51:32 2016 +1100

    Get default of TEST_SSH_UTF8 from environment.

commit b9b8ba3f9ed92c6220b58d70d1e6d8aa3eea1104
Author: Darren Tucker <dtucker@zip.com.au>
Date:   Tue Dec 13 12:56:40 2016 +1100

    Remove commented-out includes.
    
    These commented-out includes have "Still needed?" comments.  Since
    they've been commented out for ~13 years I assert that they're not.

commit 25275f1c9d5f01a0877d39444e8f90521a598ea0
Author: Darren Tucker <dtucker@zip.com.au>
Date:   Tue Dec 13 12:54:23 2016 +1100

    Add prototype for strcasestr in compat library.

commit afec07732aa2985142f3e0b9a01eb6391f523dec
Author: Darren Tucker <dtucker@zip.com.au>
Date:   Tue Dec 13 10:23:03 2016 +1100

    Add strcasestr to compat library.
    
    Fixes build on (at least) Solaris 10.

commit dda78a03af32e7994f132d923c2046e98b7c56c8
Author: Damien Miller <djm@mindrot.org>
Date:   Mon Dec 12 13:57:10 2016 +1100

    Force Turkish locales back to C/POSIX; bz#2643
    
    Turkish locales are unique in their handling of the letters 'i' and
    'I' (yes, they are different letters) and OpenSSH isn't remotely
    prepared to deal with that. For now, the best we can do is to force
    OpenSSH to use the C/POSIX locale and try to preserve the UTF-8
    encoding if possible.
    
    ok dtucker@

commit c35995048f41239fc8895aadc3374c5f75180554
Author: Darren Tucker <dtucker@zip.com.au>

openssh-7.4p1/ssh-keysign.c
/* $OpenBSD: ssh-keysign.c,v 1.52 2016/02/15 09:47:49 dtucker Exp $ */
/*
 * Copyright (c) 2002 Markus Friedl.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "includes.h"

#include <fcntl.h>
#ifdef HAVE_PATHS_H
#include <paths.h>
#endif
#include <pwd.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#ifdef WITH_OPENSSL
#include <openssl/evp.h>
#include <openssl/rand.h>
#include <openssl/rsa.h>
#endif

#include "xmalloc.h"
#include "log.h"
#include "sshkey.h"
#include "ssh.h"
#include "ssh2.h"
#include "misc.h"
#include "sshbuf.h"
#include "authfile.h"
#include "msg.h"
#include "canohost.h"
#include "pathnames.h"
#include "readconf.h"
#include "uidswap.h"
#include "sshkey.h"
#include "ssherr.h"

struct ssh *active_state = NULL; /* XXX needed for linking */

extern char *__progname;

/* XXX readconf.c needs these */
uid_t original_real_uid;

extern char *__progname;

static int
valid_request(struct passwd *pw, char *host, struct sshkey **ret,
    u_char *data, size_t datalen)
{
	struct sshbuf *b;
	struct sshkey *key = NULL;
	u_char type, *pkblob;
	char *p;
	size_t blen, len;
	char *pkalg, *luser;
	int r, pktype, fail;

	if (ret != NULL)
		*ret = NULL;
	fail = 0;

	if ((b = sshbuf_from(data, datalen)) == NULL)
		fatal("%s: sshbuf_from failed", __func__);

	/* session id, currently limited to SHA1 (20 bytes) or SHA256 (32) */
	if ((r = sshbuf_get_string(b, NULL, &len)) != 0)
		fatal("%s: buffer error: %s", __func__, ssh_err(r));
	if (len != 20 && len != 32)
		fail++;

	if ((r = sshbuf_get_u8(b, &type)) != 0)
		fatal("%s: buffer error: %s", __func__, ssh_err(r));
	if (type != SSH2_MSG_USERAUTH_REQUEST)
		fail++;

	/* server user */
	if ((r = sshbuf_skip_string(b)) != 0)
		fatal("%s: buffer error: %s", __func__, ssh_err(r));

	/* service */
	if ((r = sshbuf_get_cstring(b, &p, NULL)) != 0)
		fatal("%s: buffer error: %s", __func__, ssh_err(r));
	if (strcmp("ssh-connection", p) != 0)
		fail++;
	free(p);

	/* method */
	if ((r = sshbuf_get_cstring(b, &p, NULL)) != 0)
		fatal("%s: buffer error: %s", __func__, ssh_err(r));
	if (strcmp("hostbased", p) != 0)
		fail++;
	free(p);

	/* pubkey */
	if ((r = sshbuf_get_cstring(b, &pkalg, NULL)) != 0 ||
	    (r = sshbuf_get_string(b, &pkblob, &blen)) != 0)
		fatal("%s: buffer error: %s", __func__, ssh_err(r));

	pktype = sshkey_type_from_name(pkalg);
	if (pktype == KEY_UNSPEC)
		fail++;
	else if ((r = sshkey_from_blob(pkblob, blen, &key)) != 0) {
		error("%s: bad key blob: %s", __func__, ssh_err(r));
		fail++;
	} else if (key->type != pktype)
		fail++;
	free(pkalg);
	free(pkblob);

	/* client host name, handle trailing dot */
	if ((r = sshbuf_get_cstring(b, &p, &len)) != 0)
		fatal("%s: buffer error: %s", __func__, ssh_err(r));
	debug2("%s: check expect chost %s got %s", __func__, host, p);
	if (strlen(host) != len - 1)
		fail++;
	else if (p[len - 1] != '.')
		fail++;
	else if (strncasecmp(host, p, len - 1) != 0)
		fail++;
	free(p);

	/* local user */
	if ((r = sshbuf_get_cstring(b, &luser, NULL)) != 0)
		fatal("%s: buffer error: %s", __func__, ssh_err(r));

	if (strcmp(pw->pw_name, luser) != 0)
		fail++;
	free(luser);

	/* end of message */
	if (sshbuf_len(b) != 0)
		fail++;
	sshbuf_free(b);

	debug3("%s: fail %d", __func__, fail);

	if (fail && key != NULL)
		sshkey_free(key);
	else if (ret != NULL)
		*ret = key;

	return (fail ? -1 : 0);
}

int
main(int argc, char **argv)
{
	struct sshbuf *b;
	Options options;
#define NUM_KEYTYPES 4
	struct sshkey *keys[NUM_KEYTYPES], *key = NULL;
	struct passwd *pw;
	int r, key_fd[NUM_KEYTYPES], i, found, version = 2, fd;
	u_char *signature, *data, rver;
	char *host, *fp;
	size_t slen, dlen;
#ifdef WITH_OPENSSL
	u_int32_t rnd[256];
#endif

	ssh_malloc_init();	/* must be called before any mallocs */
	if (pledge("stdio rpath getpw dns id", NULL) != 0)
		fatal("%s: pledge: %s", __progname, strerror(errno));

	/* Ensure that stdin and stdout are connected */
	if ((fd = open(_PATH_DEVNULL, O_RDWR)) < 2)
		exit(1);
	/* Leave /dev/null fd iff it is attached to stderr */
	if (fd > 2)
		close(fd);

	i = 0;
	/* XXX This really needs to read sshd_config for the paths */
	key_fd[i++] = open(_PATH_HOST_DSA_KEY_FILE, O_RDONLY);
	key_fd[i++] = open(_PATH_HOST_ECDSA_KEY_FILE, O_RDONLY);
	key_fd[i++] = open(_PATH_HOST_ED25519_KEY_FILE, O_RDONLY);
	key_fd[i++] = open(_PATH_HOST_RSA_KEY_FILE, O_RDONLY);

	original_real_uid = getuid();	/* XXX readconf.c needs this */
	if ((pw = getpwuid(original_real_uid)) == NULL)
		fatal("getpwuid failed");
	pw = pwcopy(pw);

	permanently_set_uid(pw);

	seed_rng();

#ifdef DEBUG_SSH_KEYSIGN
	log_init("ssh-keysign", SYSLOG_LEVEL_DEBUG3, SYSLOG_FACILITY_AUTH, 0);
#endif

	/* verify that ssh-keysign is enabled by the admin */
	initialize_options(&options);
	(void)read_config_file(_PATH_HOST_CONFIG_FILE, pw, "", "", &options, 0);
	fill_default_options(&options);
	if (options.enable_ssh_keysign != 1)
		fatal("ssh-keysign not enabled in %s",
		    _PATH_HOST_CONFIG_FILE);

	for (i = found = 0; i < NUM_KEYTYPES; i++) {
		if (key_fd[i] != -1)
			found = 1;
	}
	if (found == 0)
		fatal("could not open any host key");

#ifdef WITH_OPENSSL
	OpenSSL_add_all_algorithms();
	arc4random_buf(rnd, sizeof(rnd));
	RAND_seed(rnd, sizeof(rnd));
#endif

	found = 0;
	for (i = 0; i < NUM_KEYTYPES; i++) {
		keys[i] = NULL;
		if (key_fd[i] == -1)
			continue;
		r = sshkey_load_private_type_fd(key_fd[i], KEY_UNSPEC,
		    NULL, &key, NULL);
		close(key_fd[i]);
		if (r != 0)
			debug("parse key %d: %s", i, ssh_err(r));
		else if (key != NULL) {
			keys[i] = key;
			found = 1;
		}
	}
	if (!found)
		fatal("no hostkey found");

	if (pledge("stdio dns", NULL) != 0)
		fatal("%s: pledge: %s", __progname, strerror(errno));

	if ((b = sshbuf_new()) == NULL)
		fatal("%s: sshbuf_new failed", __progname);
	if (ssh_msg_recv(STDIN_FILENO, b) < 0)
		fatal("ssh_msg_recv failed");
	if ((r = sshbuf_get_u8(b, &rver)) != 0)
		fatal("%s: buffer error: %s", __progname, ssh_err(r));
	if (rver != version)
		fatal("bad version: received %d, expected %d", rver, version);
	if ((r = sshbuf_get_u32(b, (u_int *)&fd)) != 0)
		fatal("%s: buffer error: %s", __progname, ssh_err(r));
	if (fd < 0 || fd == STDIN_FILENO || fd == STDOUT_FILENO)
		fatal("bad fd");
	if ((host = get_local_name(fd)) == NULL)
		fatal("cannot get local name for fd");

	if ((r = sshbuf_get_string(b, &data, &dlen)) != 0)
		fatal("%s: buffer error: %s", __progname, ssh_err(r));
	if (valid_request(pw, host, &key, data, dlen) < 0)
		fatal("not a valid request");
	free(host);

	found = 0;
	for (i = 0; i < NUM_KEYTYPES; i++) {
		if (keys[i] != NULL &&
		    sshkey_equal_public(key, keys[i])) {
			found = 1;
			break;
		}
	}
	if (!found) {
		if ((fp = sshkey_fingerprint(key, options.fingerprint_hash,
		    SSH_FP_DEFAULT)) == NULL)
			fatal("%s: sshkey_fingerprint failed", __progname);
		fatal("no matching hostkey found for key %s %s",
		    sshkey_type(key), fp ? fp : "");
	}

	if ((r = sshkey_sign(keys[i], &signature, &slen, data, dlen, NULL, 0))
	    != 0)
		fatal("sshkey_sign failed: %s", ssh_err(r));
	free(data);

	/* send reply */
	sshbuf_reset(b);
	if ((r = sshbuf_put_string(b, signature, slen)) != 0)
		fatal("%s: buffer error: %s", __progname, ssh_err(r));
	if (ssh_msg_send(STDOUT_FILENO, version, b) == -1)
		fatal("ssh_msg_send failed");

	return (0);
}
